File: packages/python/elizaos/bootstrap/actions/follow_room.py
"""
FOLLOW_ROOM Action - Follow a room to receive updates.

This action allows the agent to follow a room and actively
monitor messages and activities in that room.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content, MemoryType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class FollowRoomAction:
    """
    Action for following a room.

    This action is used when:
    - The agent wants to monitor a room
    - The agent should receive updates from a room
    - Active engagement with a room is desired
    """

    name: str = "FOLLOW_ROOM"
    similes: list[str] = field(
        default_factory=lambda: [
            "JOIN_ROOM",
            "SUBSCRIBE_ROOM",
            "WATCH_ROOM",
            "ENTER_ROOM",
        ]
    )
    description: str = (
        "Follow a room to receive updates and monitor messages. "
        "Use this when you want to actively engage with a room's content."
    )

    async def validate(self, runtime: IAgentRuntime, message: Memory) -> bool:
        """Validate that room information is available."""
        room_id = message.room_id
        if not room_id:
            return False

        # Check if not already following
        room = await runtime.get_room(room_id)
        if room is None:
            return False

        # Check world settings if applicable
        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                followed_rooms = world.metadata.get("followedRooms", [])
                # Already following - no need to follow again
                if str(room_id) in followed_rooms:
                    return False

        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        """Handle following a room."""
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to follow",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "FOLLOW_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "FOLLOW_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        try:
            # Get world and update followed rooms
            world_id = room.world_id
            if world_id:
                world = await runtime.get_world(world_id)
                if world and world.metadata:
                    followed_rooms = list(world.metadata.get("followedRooms", []))
                    room_id_str = str(room_id)

                    if room_id_str not in followed_rooms:
                        followed_rooms.append(room_id_str)
                        world.metadata["followedRooms"] = followed_rooms
                        await runtime.update_world(world)

            # Create memory of the action
            await runtime.create_memory(
                content=Content(
                    text=f"Now following room: {room_name}",
                    actions=["FOLLOW_ROOM"],
                ),
                room_id=room_id,
                entity_id=runtime.agent_id,
                memory_type=MemoryType.ACTION,
                metadata={"type": "FOLLOW_ROOM", "roomName": room_name},
            )

            response_content = Content(
                text=f"I am now following {room_name} and will monitor its messages.",
                actions=["FOLLOW_ROOM"],
            )

            if callback:
                await callback(response_content)

            return ActionResult(
                text=f"Now following room: {room_name}",
                values={
                    "success": True,
                    "following": True,
                    "roomId": str(room_id),
                    "roomName": room_name,
                },
                data={
                    "actionName": "FOLLOW_ROOM",
                    "roomId": str(room_id),
                    "roomName": room_name,
                },
                success=True,
            )

        except Exception as error:
            runtime.logger.error(
                {
                    "src": "plugin:bootstrap:action:followRoom",
                    "agentId": runtime.agent_id,
                    "roomId": str(room_id),
                    "error": str(error),
                },
                "Error following room",
            )
            return ActionResult(
                text="Error following room",
                values={"success": False, "error": str(error)},
                data={"actionName": "FOLLOW_ROOM", "error": str(error)},
                success=False,
                error=error,
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        """Example interactions demonstrating the FOLLOW_ROOM action."""
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Can you keep an eye on this channel?"),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll follow this room and monitor its activity.",
                        actions=["FOLLOW_ROOM"],
                    ),
                ),
            ],
        ]


# Create the action instance
follow_room_action = Action(
    name=FollowRoomAction.name,
    similes=FollowRoomAction().similes,
    description=FollowRoomAction.description,
    validate=FollowRoomAction().validate,
    handler=FollowRoomAction().handler,
    examples=FollowRoomAction().examples,
)

File: packages/rust/src/bootstrap/actions/follow_room.rs
//! FOLLOW_ROOM action implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

use crate::error::{PluginError, PluginResult};
use crate::runtime::IAgentRuntime;
use crate::types::{ActionResult, Content, Memory, MemoryType, State};

use super::Action;

/// Action for following a room.
pub struct FollowRoomAction;

#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
impl Action for FollowRoomAction {
    fn name(&self) -> &'static str {
        "FOLLOW_ROOM"
    }

    fn similes(&self) -> &[&'static str] {
        &["JOIN_ROOM", "SUBSCRIBE_ROOM", "WATCH_ROOM", "ENTER_ROOM"]
    }

    fn description(&self) -> &'static str {
        "Follow a room to receive updates and monitor messages. \
         Use this when you want to actively engage with a room's content."
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        let Some(room_id) = message.room_id else {
            return false;
        };

        // Check if room exists
        let Ok(Some(room)) = runtime.get_room(room_id).await else {
            return false;
        };

        // Check if already following
        if let Some(world_id) = room.world_id {
            if let Ok(Some(world)) = runtime.get_world(world_id).await {
                if let Some(followed) = world.metadata.get("followedRooms") {
                    if let Some(arr) = followed.as_array() {
                        let room_str = room_id.to_string();
                        if arr.iter().any(|v| v.as_str() == Some(&room_str)) {
                            return false; // Already following
                        }
                    }
                }
            }
        }

        true
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        _state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let room_id = message.room_id.ok_or_else(|| {
            PluginError::InvalidInput("No room specified to follow".to_string())
        })?;

        let room = runtime
            .get_room(room_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Room not found".to_string()))?;

        let room_name = room.name.clone().unwrap_or_else(|| "Unknown Room".to_string());

        // Update world's followed rooms
        if let Some(world_id) = room.world_id {
            if let Some(mut world) = runtime.get_world(world_id).await? {
                let mut followed: Vec<String> = world
                    .metadata
                    .get("followedRooms")
                    .and_then(|v| v.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|v| v.as_str().map(String::from))
                            .collect()
                    })
                    .unwrap_or_default();

                let room_str = room_id.to_string();
                if !followed.contains(&room_str) {
                    followed.push(room_str);
                    world.metadata.insert(
                        "followedRooms".to_string(),
                        serde_json::json!(followed),
                    );
                    runtime.update_world(&world).await?;
                }
            }
        }

        // Create memory of the action
        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("FOLLOW_ROOM"));
        metadata.insert("roomName".to_string(), serde_json::json!(&room_name));

        runtime
            .create_memory(
                Content {
                    text: format!("Now following room: {}", room_name),
                    actions: vec!["FOLLOW_ROOM".to_string()],
                    ..Default::default()
                },
                Some(room_id),
                Some(runtime.agent_id()),
                MemoryType::Action,
                metadata,
            )
            .await?;

        Ok(ActionResult::success(format!("Now following room: {}", room_name))
            .with_value("success", true)
            .with_value("following", true)
            .with_value("roomId", room_id.to_string())
            .with_value("roomName", room_name.clone())
            .with_data("actionName", "FOLLOW_ROOM")
            .with_data("roomId", room_id.to_string())
            .with_data("roomName", room_name))
    }
}

